rsocket-core/src/main/java/io/rsocket/internal/FluxSwitchOnFirst.java

```diff
@@ -0,0 +1,657 @@
+/*
+ * Copyright (c) 2011-2018 Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.rsocket.internal;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.BiFunction;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+import reactor.core.Fuseable;
+import reactor.core.Scannable;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxOperator;
+import reactor.core.publisher.Operators;
+import reactor.core.publisher.Signal;
+import reactor.util.annotation.Nullable;
+import reactor.util.context.Context;
+
+/**
+ * @author Oleh Dokuka
+ * @param <T>
+ * @param <R>
+ */
+public final class FluxSwitchOnFirst<T, R> extends FluxOperator<T, R> {
```

Comment:
this should at least very explicitly document the differences with the current core implementation and why this was (re)introduced by this PR when there already is an operator in core... and at least link to the PR/issue in core (if it is as relevant as I presume)

Reply:
In general, this implementation exposes the intent I want to mirror in the switchOnFirst in the reactor. However, this code will be sitting there until I'm going to write custom operators for all the communication types. I want to do something similar to what is done in rsocket-cpp -> https://github.com/rsocket/rsocket-cpp/tree/master/rsocket/statemachine

Reply:
Illuminated. Using the fixed one from the latest Reactor-Core